### DISCUSSION TITLE
Panic: assertion failed in BasicNftTest.t.sol

### DISCUSSION THREAD

**Discussion by monteago30** [Cyfrin/foundry-full-course-cu#1459] [General]
I am getting Panic: assertion failed (0x01) after running testCanMintAndHaveABalance() test. I believe the PUG hash is not equal to the tokenURI hash.  How can i fix our verify the tokenURI hash?

BasicNftTest.t.sol:
```solidity
contract BasicNftTest is Test {
    DeployBasicNft public deployer;
    BasicNft public basicNft;
    address public USER = makeAddr("user");
    string public constant PUG = "ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json";

function testCanMintAndHaveABalance() public {
        vm.prank(USER);
        basicNft.mintNft(PUG);

        assert(basicNft.balanceOf(USER) == 1);
        assert(keccak256(abi.encodePacked(PUG)) == keccak256(abi.encodePacked(basicNft.tokenURI)));
    }
```

BasicNft.sol:
```
function mintNft(string memory tokenUri) public {
        s_tokenIdToUri[s_tokenCounter] = tokenUri;
        _safeMint(msg.sender, s_tokenCounter);
        s_tokenCounter++;
    }

    function tokenURI(uint256 tokenId) public view override returns (string memory) {
        return s_tokenIdToUri[tokenId];
    } 
```

Ran 1 test for test/BasicNftTest.t.sol:BasicNftTest
[FAIL. Reason: panic: assertion failed (0x01)] testCanMintAndHaveABalance() (gas: 154874)
Traces:
  [154874] BasicNftTest::testCanMintAndHaveABalance()
    ├─ [0] VM::prank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← ()
    ├─ [142158] BasicNft::mintNft("ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json")
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], tokenId: 1)
    │   └─ ← ()
    ├─ [612] BasicNft::balanceOf(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D]) [staticcall]
    │   └─ ← 1
    └─ ← panic: assertion failed (0x01)

**Comment by EngrPips** [comment]
Hello @monteago30,  according to what you show above `tokenURI` on `BasicNft .sol` takes in an argument of type uint256 named `tokenId` to return the `URI` corresponding to the passed in `tokenId` but when you tried to read the `URI` from the `BasicNft .sol` in your test file you didnt't pass in any argument to the `tokenURI` function, why is that ? 

You have the below in your test file 

```javascript
assert(keccak256(abi.encodePacked(PUG)) == keccak256(abi.encodePacked(basicNft.tokenURI)));
```

While you are suppose to actually have 

```javascript
assert(keccak256(abi.encodePacked(PUG)) == keccak256(abi.encodePacked(basicNft.tokenURI(tokenId))));
```

So i would say make sure to know what is the tokenId of the pug nft you minted in `testCanMintAndHaveABalance ` test as shown in your code above 

```javscript
 basicNft.mintNft(PUG);
```

And pass the tokenId in to the function to read from the `BasicNft.sol` contract.

```javscript
keccak256(abi.encodePacked(basicNft.tokenURI(tokenId))
```

Try these and let's see if the error still persist.

**Comment by monteago30** [comment]
Hi @EngrPips - thank you for your response and some guidance.  Passing in the (tokenId) didn't actually work.  I get an undeclared variable when I add that.

```solidity
 assert(keccak256(abi.encodePacked(PUG)) == keccak256(abi.encodePacked(basicNft.tokenURI(tokenId)))
 ```
 Compiler run failed:
Error (7576): Undeclared identifier.
  --> test/BasicNftTest.t.sol:33:97:
   |
33 |         assert(keccak256(abi.encodePacked(PUG)) == keccak256(abi.encodePacked(basicNft.tokenURI(tokenId)))
   |                                                                                                 ^^^^^^^^
   
   However you made me look at the tutorial again and realizing that i was supposed to pass in (0):
   ```solidity
   assert(keccak256(abi.encodePacked(PUG)) == keccak256(abi.encodePacked(basicNft.tokenURI(0)))
   ```
   
   but that still results in the Panic: assertion failure...when i change it to (1) it works:
   
   ```solidity
   assert(keccak256(abi.encodePacked(PUG)) == keccak256(abi.encodePacked(basicNft.tokenURI(1)))
   ```
   
   
   
![BasicNft sol](https://github.com/Cyfrin/foundry-full-course-f23/assets/30839299/1e554e03-a090-4ceb-a433-c81945f3acb0)

**Reply by EngrPips** [reply]
I am glad you one way or the other realize what i was pointing you to.  You can mark this discussion as answered if you think your problem have been resolved. thanks @monteago30

**Reply by monteago30** [reply]
Hi @EngrPips thank you for the help...I think i am good, I can move on and mark this as closed.  I am wondering why it was expecting a value of (1) rather then (0) as described in the tutorial.  Thanks!

**Reply by EngrPips** [reply]
You are welcome @monteago30 , it was expecting a value of (1) because you started the `tokenId` from (1) and not (0) , You can check your `BasicNft .sol` to confirm that.

**Reply by monteago30** [reply]
@EngrPips you are correct when i changed the s_tokenCounter to (0) it worked with by passing (0) into the test function.

```solidity
contract BasicNft is ERC721 {
    uint256 private s_tokenCounter;
    mapping(uint256 => string) private s_tokenIdToUri;

    constructor() ERC721("Dogie", "DOG") {
        s_tokenCounter = 0;
    }
    ```

**Reply by monteago30** [reply]
good deal I'm learning...thank you!!

**Reply by EngrPips** [reply]
I am glad to help friend, but generally it good to starts `Ids` and other stuff from (1) instead of (0) because you know default value for any type of uint or int in solidity is (0). so you don't usually want to start from (0). @monteago30
